roles/tests-integration/tests/common/proxy/mod.rs

```diff
@@ -0,0 +1,489 @@
+                    send_original_frame_upstream = false;
+                    if send.send(frame).await.is_err() {
+                        return Err(ProxyError::UpstreamClosed);
+                    };
+                }
+            }
+            if send_original_frame_upstream && send.send(frame).await.is_err() {
+                return Err(ProxyError::UpstreamClosed);
+            };
+        }
+        Err(ProxyError::DownstreamClosed)
+    }
+
+    async fn recv_from_up_send_to_down(
+        mut recv: Receiver<MessageFrame>,
+        send: Sender<MessageFrame>,
+        mut handlers: Vec<MessageChannel>,
+    ) -> Result<(), ProxyError> {
+        while let Some(mut frame) = recv.recv().await {
+            let mut send_original_frame_upstream = true;
+            for handler in handlers.iter_mut() {
+                if let Some(frame) = handler.on_message(&mut frame).await {
+                    send_original_frame_upstream = false;
+                    if send.send(frame).await.is_err() {
+                        return Err(ProxyError::DownstreamClosed);
+                    };
+                }
+            }
+            if send_original_frame_upstream && send.send(frame).await.is_err() {
+                return Err(ProxyError::DownstreamClosed);
+            };
+        }
+        Err(ProxyError::UpstreamClosed)
+    }
+}
+
+#[derive(PartialEq, Clone, Debug)]
+pub enum Role {
+    Client,
+    Server,
+}
+
+impl std::fmt::Display for Role {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Role::Client => write!(f, "Client"),
+            Role::Server => write!(f, "Server"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct MessagesAggregator {
+    messages: Arc<Mutex<Vec<(PoolMessages<'static>, Role)>>>,
+}
+
+impl MessagesAggregator {
+    pub fn new() -> Self {
+        Self {
+            messages: Arc::new(Mutex::new(vec![])),
+        }
+    }
+
+    pub fn add_message(&self, message: PoolMessages<'static>, role: Role) {
+        self.messages
+            .safe_lock(|messages| {
+                let mut new_state = messages.clone();
+                new_state.push((message, role));
+                *messages = new_state
+            })
+            .unwrap();
+    }
+
+    pub fn contains(&self, message: ExpectMessage, role: Role) -> bool {
+        let messages = self
+            .messages
+            .safe_lock(|messages| messages.clone())
+            .unwrap();
+        let messages = messages
+            .iter()
+            .map(|(m, r)| (ExpectMessage::from(m.clone()), r.clone()));
+        for m in messages {
+            if m.0 == message && m.1 == role {
+                return true;
+            }
+        }
+        false
+    }
+}
+
+struct MessageChannel {
+    pub message_type: u8,
+    pub expect_from: Role,
+    pub receiver: Option<Receiver<PoolMessages<'static>>>,
+    pub sender: Sender<PoolMessages<'static>>,
+}
```

Comment:
the concept of `handler` and this struct `MessageChannel` are not clear to me

what is their purpose?

Reply:
I see this is being inherited from `demand_easy_sv2`

but I wonder if we really need it?